experimental/devices/nrzled/nrzled.go

```diff
@@ -0,0 +1,221 @@
+type Dev struct {
+	p         gpiostream.PinOut
+	numPixels int
+	channels  int                  // Number of channels per pixel
+	b         gpiostream.BitStream // NRZ encoded bits; cached to reduce heap fragmentation
+	buf       []byte               // Double buffer of RGB/RGBW pixels; enables partial Draw()
+}
+
+func (d *Dev) String() string {
+	return fmt.Sprintf("nrzled{%s}", d.p)
+}
+
+// Halt turns the lights off.
+func (d *Dev) Halt() error {
+	// It doesn't affect the back buffer.
+	zero := NRZ(0)
+	a := byte(zero >> 16)
+	b := byte(zero >> 8)
+	c := byte(zero)
+	for i := 0; i < d.channels*d.numPixels; i++ {
+		d.b.Bits[3*i+0] = a
+		d.b.Bits[3*i+1] = b
+		d.b.Bits[3*i+2] = c
+	}
+	if err := d.p.StreamOut(&d.b); err != nil {
+		return fmt.Errorf("nrzled: %v", err)
+	}
+	return nil
+}
+
+// ColorModel implements devices.Display.
+//
+// It is color.NRGBAModel.
+func (d *Dev) ColorModel() color.Model {
+	return color.NRGBAModel
+}
+
+// Bounds implements devices.Display. Min is guaranteed to be {0, 0}.
+func (d *Dev) Bounds() image.Rectangle {
+	return image.Rectangle{Max: image.Point{X: d.numPixels, Y: 1}}
+}
+
+// Draw implements devices.Display.
+//
+// Using something else than image.NRGBA is 10x slower and is not recommended.
+// When using image.NRGBA, the alpha channel is ignored in RGB mode and used as
+// White channel in RGBW mode.
+//
+// A back buffer is kept so that partial updates are supported, albeit the full
+// LED strip is updated synchronously.
+func (d *Dev) Draw(r image.Rectangle, src image.Image, sp image.Point) {
+	r = r.Intersect(d.Bounds())
+	srcR := src.Bounds()
+	srcR.Min = srcR.Min.Add(sp)
+	if dX := r.Dx(); dX < srcR.Dx() {
+		srcR.Max.X = srcR.Min.X + dX
+	}
+	if dY := r.Dy(); dY < srcR.Dy() {
+		srcR.Max.Y = srcR.Min.Y + dY
+	}
+	if d.buf == nil {
+		// Allocate d.buf on first Draw() call, in case the user only wants to use
+		// .Write().
+		d.buf = make([]byte, d.numPixels*d.channels)
+	}
+	if img, ok := src.(*image.NRGBA); ok {
+		// Fast path for image.NRGBA.
+		base := srcR.Min.Y * img.Stride
+		raster(d.b.Bits, img.Pix[base+4*srcR.Min.X:base+4*srcR.Max.X], d.channels, 4)
+	} else {
+		// Generic version.
+		m := srcR.Max.X - srcR.Min.X
+		if d.channels == 3 {
+			for i := 0; i < m; i++ {
+				c := color.NRGBAModel.Convert(src.At(srcR.Min.X+i, srcR.Min.Y)).(color.NRGBA)
+				j := 3 * i
+				put(d.b.Bits[3*(j+0):], c.G)
+				put(d.b.Bits[3*(j+1):], c.R)
+				put(d.b.Bits[3*(j+2):], c.B)
+			}
+		} else {
+			for i := 0; i < m; i++ {
+				c := color.NRGBAModel.Convert(src.At(srcR.Min.X+i, srcR.Min.Y)).(color.NRGBA)
+				j := 4 * i
+				put(d.b.Bits[3*(j+0):], c.G)
+				put(d.b.Bits[3*(j+1):], c.R)
+				put(d.b.Bits[3*(j+2):], c.B)
+				put(d.b.Bits[3*(j+3):], c.A)
+			}
+		}
+	}
+	_ = d.p.StreamOut(&d.b)
+}
+
+// Write accepts a stream of raw RGB/RGBW pixels and sends it as NRZ encoded
+// stream.
+//
+// This bypasses the back buffer.
+func (d *Dev) Write(pixels []byte) (int, error) {
```

Comment:
I think this API can be better documented. I expected it to be the defacto API to set the pixels but your comment about bypassing the back buffer makes it sound like this is more of a hack or workaround to a better solution. Is it?
Also I'm not sure what to set the byte values to, is that a RGB byte value?